### DISCUSSION TITLE
Is there a way to return the query given a index matrix?

### DISCUSSION THREAD

**Discussion by MarshalLeeeeee** [dimforge/nalgebra#1381] [Q&A] [answered]
I am doing a simulation project with the case of advection of the grid-based density field given the velocity. One common way is to perform an inverse advection, and to interpolate densities around. This algorithm basically performs something like below on every grid, where (i,j) is the index of the grid.
```
// backward advection
let ii = i - vx*dt;
let jj = j - vy*dt;

// get the densities around
let x00 = m.index((floor(ii), floor(jj)));
let x01 = m.index((floor(ii), ceil(jj)));
let x10 = m.index((ceil(ii), floor(jj)));
let x11 = m.index((ceil(ii), ceil(jj)));
// ... perform interpolation ...
// ... update m(i,j) with interpolated value ...
```

So I'm wondering if there is an api ```xxx```, to take ```I``` and ```J```, and return the matrix sharing the same size as ```I``` and ```J```. Like
```
let m_indexed = m.xxx(I, J);
'''

If there is an existing one, please tell me. 
Or any equivalent solution with existing apis are welcomed to be shared.
Thanks.

**Comment by Andlon** [comment]
Hi there. I don't understand your question - what are `I` and `J`, and what is "the matrix sharing the same size as `I` and `J`"?

**Reply by MarshalLeeeeee** [reply]
Thanks for ur reply and apologize for my ambiguity. Actually, ```I``` and ```J``` does not have to be the same size with the query matrix ```M```.
What I want to say is like the toy case below, suppose
```
M = [
    1, 2, 3,
    4, 5, 6,
    7, 8, 9,
]
I = [
    0, 1,
    1, 2, 
]
J = [
    2, 1,
    0, 2,
]
```
Then is there any api can return the query ```Q```, such that ```Q[[i, j]] = M[[I[[i,j]], J[[i,j]]]]```, where in the above example, ```Q``` should be
```
Q = [
    3, 5,
    4, 9,
]
```

**Reply by Ralith** [reply, marked as answer]
You want [`Matrix::zip_map`](https://docs.rs/nalgebra/latest/nalgebra/base/struct.Matrix.html#method.zip_map), with something like `I.zip_map(&J, |x, y| M[(x, y)])`.

**Reply by MarshalLeeeeee** [reply]
Thanks a lot, it works!
